server-function/src/index.ts

```diff
@@ -0,0 +1,57 @@
+import path from 'path';
+import fs from 'mz/fs';
+import mimeTypes from 'mime-types';
+
+const UP_PATH_REGEXP = /(?:^|[\\/])\.\.(?:[\\/]|$)/;
+
+export async function isFile(filePath: string): Promise<boolean> {
+  try {
+    const stats = await fs.stat(filePath);
+    return stats.isFile();
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      return false;
+    }
+    throw err;
+  }
+}
+
+export class Server {
+  constructor(private directory: string, private extensions: string[] = ['.html']) {
+  }
+  public async handle(url: string): Promise<any> {
+    if (url === '/invoke') {
+      return { action: 'handleInvoke' };
+    }
+    if (url === '/') {
+      return this.handle('/index.html');
+    }
+    const fullPath = path.join(this.directory, url);
+    if (UP_PATH_REGEXP.test(fullPath)) {
+      return { action: 'sendStatus', status: 403 };
```

Comment:
most sites handle this by treating `..` of the root as `/` and would return 404 as usual if required.
for example: `../index.html` would just work if index.html exists.
i think you can mimic this with `path.join(this.directory, path.join('/', url))` and avoid the regexp completely